### PR TITLE
Trinity.C: decouple event handling from networking.

### PR DESCRIPTION
**Pull request by yatli**
Decoupling event handling from networking is surprisingly easy -- the two parts are almost orthogonal.

This gives us a chance to repurpose the worker thread pool -- currently only for server-side networking I/O:

- We could use the thread pool to handle client IO asynchronously. When a client IO finishes, use `worktype_t` to indicate that we want to wake up a sleepy client.
- We could use the thread pool to host continuations. a message handler can then return a watch rather than a response. The typical use case is to prevent recursive SYNC_WITH_RSP messages draining the workers/client connections.
  - get a client -> Server handler blocked -> get a client -> another server handler blocked -> ...
  - vs: Server handler blocked -> get a client, escape with `co_await` -> eventloop calls co_resume -> ...
- With the networking decoupled, the user can queue custom work items into the pool. Continuations still apply here.

### CODE REVIEW

**Comment by yatli** [comment]
further down the path, the decoupling of platform-specific send/recv primitives from the message handling workflow in turn unified the networking loop of iocp/unix sides.

We are then able to easily implement a true async networking loop with epoll.

the key difference between IOCP and epoll/kqueue is that, for IOCP, `send_async` is a true call to send, but for posix we just rearm the FD to listen for writable events. we then differ the actual call to recv to `poll_work`, where `_eventloop` expects bytes to be pumped from a socket.

**Comment by yatli** [comment]
... and that’s it, iocp/epoll done. This time it’s true async.

**Comment by TaviTruman** [comment]
Very exciting news. Can't wait to start leveraging this capability

**Comment by yatli** [comment]
Blocking issue: https://github.com/Microsoft/GraphEngine/issues/220
Edit: closed

File: src/Trinity.C/src/Events/Events.cpp [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                switch (pwork->type)
+                {
+                case worktype_t::Shutdown:
+                    free_work(pwork);
+                    return;
+                case worktype_t::Receive:
+
+                    if (!Network::process_recv(psock, szwork))
+                    {
+                        break;
```
Comment (yatli) [inline src/Trinity.C/src/Events/Events.cpp:57]: need to rearm and call recv_async here?

Comment (yatli) [reply src/Trinity.C/src/Events/Events.cpp:57]: no, rearmed in process_recv.

**Review by shaobin — APPROVED** [review]
looks good to me.

**Pull request merged**